exotica_core/src/task_map.cpp

```diff
@@ -73,24 +73,36 @@ void TaskMap::Update(Eigen::VectorXdRefConst x, Eigen::VectorXdRef Phi, Eigen::M
     if (jacobian.rows() != TaskSpaceDim() && jacobian.cols() != x.rows())
         ThrowNamed("Jacobian dimension mismatch!");
 
-    // Store original x (needs to be reset later).
-    Eigen::VectorXd x_original(x);
-    Update(x_original, Phi);
-
-    // Backward finite differencing.
+    // Set constants
     constexpr double h = 1e-6;
-    Eigen::VectorXd x_tmp;
-    Eigen::VectorXd phi_original(Phi);
-    for (int i = 0; i < TaskSpaceDim(); ++i)
+    constexpr double hi = 1.0 / h;
+    int ntask = TaskSpaceDim();
+    int ndof = jacobian.cols();
+
+    // Compute x/Phi using forward mapping (no jacobain)
+    Update(x, Phi);
+
+    // Backward finite differencing
+    for (int i = 0; i < ndof; ++i)
     {
-        x_tmp = x;
-        x_tmp(i) -= h;
-        Update(x_tmp, Phi);
-        jacobian.row(i) = (1 / h) * (phi_original - Phi);
+        // Setup for gradient estimation
+        Eigen::VectorXd x_down = x;
+        x_down(i) -= h;
+        Eigen::VectorXd phi_down;
+        phi_down.resize(ntask);
+
+        // Set x_down as model state
+        scene_->GetKinematicTree().SetModelState(x_down);
```

Comment:
`SetModelState` or `Update`? I believe this should be `Update`, otherwise will break floating-base or versions where only a subset of joints is controlled.

Reply:
* I have tested both `scene_->GetKinematicTree().Update(x)` and `scene_->GetKinematicTree().SetModelState(x)` on kuka and valkyrie models
* Both `SetModelState` and `Update` work for the simplified kuka but `SetModelState` fail on valkyrie, throws an error re wrong start state size
* Will update to use `scene_->GetKinematicTree().Update(x)`

Reply:
Yep, that is expected... because SetModelState requires the full state to be set. While x is only the currently active planning group So for wherever planning group != whole-body state (e.g. only planning for a subgroup) this would fail.